### ISSUE TITLE
VFS on Linux

### ISSUE THREAD

**Issue by tobiasKaminsky** [nextcloud/desktop#3668]
VFS on Linux needs to have a proper support on KDE/Gnome first, so first step would be to create and commit such a RFC.
Then monitor it, so that we can afterwards implement it.

As many Linux users also use CLI, maybe we can also enhance our nextcloud to sync/upload/download, e.g.:
nextcloudcmd myVirtualImage.png.nextcloud -> downloads and opens it

**Comment by vasyugan** [comment]
For consistency, VFS support on Linux should not be bound to any specific desktop environment. A situation where nautilus or dolphin show some file as present yet command line utilities or even graphical programmes that don't use the GNOME/KDE libraries don't would be rather odd and irritating.

**Comment by FlexW** [comment]
@vasyugan @tobiasKaminsky At the moment it is planned to register a custom xattr and read that when opening files. That would still need support from the program that reads the file, but it would not be bound to GNOME or KDE. A real vfs like on Windows and macOS would need support from the kernel side I think.

**Comment by johnnyasantoss** [comment]
pCloud does an excellent job at using virtual fs to browse/sync files. I think that it would be a great addition to the Nextcloud Linux Client if it could do the same. 
They created their own fs it might be a good reference https://github.com/pcloudcom/pclsync

**Comment by vasyugan** [comment]
>  A real vfs like on Windows and macOS would need support from the kernel side I think.

Could this possibly be implemented using FUSE? This way, you would not have to get code into the kernel, I guess.

**Comment by mgallien** [comment]
> > A real vfs like on Windows and macOS would need support from the kernel side I think.
> 
> Couldn't such a thing be implemented using FUSE? This way, you would not have to get code into the kernel, I guess. In other words: Isn't that exactly what FUSE can do for you?

FUSE is not a good solution when network is involved because the normal file API you end up using to access FUSE filesystems is not able to cope with network errors.
Currently we have an experimental VFS feature on all platforms that is using some suffix appended to files when they are virtual empty placeholders.
You can activate it by manually editing your configuration file (`$HOME/.config/Nextcloud/nextcloud.cfg`) adding `showExperimentalOptions = true` in the first section.
You can then use the settings dialog to enable it.
Keep in mind that this experimental and not considered stable enough for daily usage.
Reports are more than welcome (bugs, performance issues, suggestions, ...).

**Comment by AditiThirdEye** [comment]
Hello @FlexW , @mgallien 
I would like to contribute to this thread, can you please suggest which approach would be the best to implement the same.
Regards,
Aditi

**Comment by AditiThirdEye** [comment]
Hello @FlexW , @mgallien ,
If the PR has not been done yet, can I work on this thread?

Regards,
Aditi

**Comment by mgallien** [comment]
> Hello @FlexW , @mgallien ,
> If the PR has not been done yet, can I work on this thread?
> 
> Regards,
> Aditi

Hello Aditi feel free to contribute by opening a PR
you can find some documentation on our workflow here [wiki](https://github.com/nextcloud/desktop/wiki)
especially the page about [Pull requests](https://github.com/nextcloud/desktop/wiki/How-to-manage-pull-requests)

**Comment by tennox** [comment]
For reference, here's a Prototype implemented for Owncloud:
https://github.com/jnweiger/OCFFS

**Comment by vasyugan** [comment]
> Currently we have an experimental VFS feature on all platforms that is using some suffix appended to files when they are virtual empty placeholders.  

I suppose that this only works with GTK/QT applications, right? I definitely would need something that also works in the shell. I very often find that I can get things done quicker and with more ease in the shell.

**Comment by KaeTuuN** [comment]
@mgallien @tobiasKaminsky 
1. Is there any (Dev) Documentation on how this feature Works on Linux?
2. Is there any Roadmap for this feature?
3. How can I support here with tests?

My employer is searching for a solution like nextcloud, but a must have is a VFS for Mac and Linux.

**Comment by tobiasKaminsky** [comment]
Currently there is no such thing for Linux.
First we implemented it on Windows and then we want to focus on MacOS, as there is a proper cloud API already.
Then we will start with Linux

**Comment by sinni800** [comment]
When I enable the experimental support that is there, it crashes when i hit the "enable virtual file support" button.
The application segfaults and the last line in the log is this:

`2022-01-08 10:10:33:653 [ warning default unknown:0 ]:  QObject::connect(QDialog, OCC::AccountSettings): invalid nullptr parameter
`

**Comment by aviallon** [comment]
Using FUSE for such a feature is clearly the right thing to do here, in my opinion.

**Comment by Cloud-Kid** [comment]
Any update on this ? I use Arch + Gnome and the client really lack of this feature :(

**Comment by vasyugan** [comment]
> Using FUSE for such a feature is clearly the right thing to do here, in my opinion.

As @mgallien writes, fuse is unable to cope with network errors, so it would need to be fixed first.

**Comment by aviallon** [comment]
> > Using FUSE for such a feature is clearly the right thing to do here, in my opinion.
> 
> As @mgallien writes, fuse is unable to cope with network errors, so it would need to be fixed first.

Correct me if I'm wrong, but I think kio-fuse is able to cope with network errors, and actually implements webdav support.
Couldn't we look at what they currently do?

**Comment by SuperSandro2000** [comment]
I did a little research and at least for dolphin/kio the xattr support does not seem to be great. Also there is no universal xattr support so many tools and programs seem to strip them on save.

ref https://userbase.kde.org/Baloo

**Comment by hardwareadictos** [comment]
Is showExperimentalOptions still available on v3.4.4? I'm on Fedora 36 but i am not able to see it under settings rear adding the line on the config file and restarting.

**Comment by hardwareadictos** [comment]
> > Is showExperimentalOptions still available on v3.4.4? I'm on Fedora 36 but i am not able to see it under settings rear adding the line on the config file and restarting.
> 
> yes should still work as we did not change anything there

![imagen](https://user-images.githubusercontent.com/20403906/170221726-6545d210-ad5f-41cd-9e0a-9cdb8babb61e.png)

¿Maybe a Fedora issue? On F35 it was working fine.

**Comment by tennox** [comment]
> ¿Maybe a Fedora issue? On F35 it was working fine.

Same for me - my solution was to switch from flatpak version to [AppImage](https://nextcloud.com/install/#install-clients)

Works now :)

**Comment by Morganlej** [comment]
To those who say it is working for you on Linux: what is working and what not?   What are the problems?

For me it is good enough if it only works in file browser, and when I click it it gets downloaded without the extra name suffix, as described [here.](https://discuss.getsol.us/d/7477-virtual-file-system-support-in-nextcloud-client)

_(Tried to enable it in the readily packaged client on Mageia, but it segfaults like described Jan 8 and just here above.  I might try Flatpak or Appimage another day)_

**Comment by tobiasKaminsky** [comment]
You need to:
1. Set showExperimentalOptions=true in the the nextcloud.cfg
2. Associate *.nextcloud files with Nextcloud client executable, so they get opened with desktop client when double-clicked
3. switch to VFS, best is to set up a new folder synchronisation

**Comment by XueSheng-GIT** [comment]
> What are the problems?

While testing on NC25, I noticed that the new feature "local editing of files" doesn't work with VFS enabled. At least if the file was not syned/downloaded yet.

**Comment by Zocker1999NET** [comment]
> fuse is unable to cope with network errors, so it would need to be fixed first.

What about the handling [sshfs](https://github.com/libfuse/sshfs) uses?

In general I would also prefer a FUSE solution as it would work natively with all applications (GUI & shell) and if implemented even allows to save bandwidth when seeking in a big file.
E.g. when an application like `file` only needs a few blocks of a file, with FUSE you could translate those seekings to their corresponding HTTP range requests ([this library](https://github.com/miquels/webdavfs) try to implement that).
The current solution wouldn't allow such improvements later without changing the whole concept.

**Comment by ACiDGRiM** [comment]
I'm using the AppImage and I created a custom mime xml for .nextcloud and associated it with the AppImage executable. but when I pass it in cmd or open with the .nextcloud file, I get "nextcloud.gui.application: Already running, exiting..."

**Comment by mgallien** [comment]
> I'm using the AppImage and I created a custom mime xml for .nextcloud and associated it with the AppImage executable. but when I pass it in cmd or open with the .nextcloud file, I get "nextcloud.gui.application: Already running, exiting..."

that is expected and it should forward the request to the already running instance

**Comment by ACiDGRiM** [comment]
> > I'm using the AppImage and I created a custom mime xml for .nextcloud and associated it with the AppImage executable. but when I pass it in cmd or open with the .nextcloud file, I get "nextcloud.gui.application: Already running, exiting..."
> 
> that is expected and it should forward the request to the already running instance

Understood, I didn't know that is how it's supposed to behave. However no file gets downloaded or indication that it's waiting for sync. Is there a debug facility?

**Comment by mgallien** [comment]
> > > I'm using the AppImage and I created a custom mime xml for .nextcloud and associated it with the AppImage executable. but when I pass it in cmd or open with the .nextcloud file, I get "nextcloud.gui.application: Already running, exiting..."
> > 
> > 
> > that is expected and it should forward the request to the already running instance
> 
> Understood, I didn't know that is how it's supposed to behave. However no file gets downloaded or indication that it's waiting for sync. Is there a debug facility?

yes
in the settings dialog, one can generate a debug archive to get an archive with the logs
or you can try reading them on your own
should be located in `~/.config/Nextcloud/logs`

**Comment by ACiDGRiM** [comment]
> > > > I'm using the AppImage and I created a custom mime xml for .nextcloud and associated it with the AppImage executable. but when I pass it in cmd or open with the .nextcloud file, I get "nextcloud.gui.application: Already running, exiting..."
> > > 
> > > 
> > > that is expected and it should forward the request to the already running instance
> > 
> > 
> > Understood, I didn't know that is how it's supposed to behave. However no file gets downloaded or indication that it's waiting for sync. Is there a debug facility?
> 
> yes in the settings dialog, one can generate a debug archive to get an archive with the logs or you can try reading them on your own should be located in `~/.config/Nextcloud/logs`

I see what the issue is...I have symbolic links to Desktop,Documents,Downloads,etc... in my home directory which allow me to right click and choose the Nextcloud options "make available online, share, etc..." but because the file's symbolic path is not in "Nextcloud" folder, it's ignored by the VFS handler.

It would be convienent if the .nextcloud files stored an obfuscated file identifier within them, so that the desktop client relied on that instead of the path of the file.

I can reproduce this issue by changing the directory directly to where the file is stored, and running 
`nextcloud.AppImage myfile.pdf.nextcloud`
This fails to launch the file, whereas 
`nextcloud.AppImage /home/user/Nextcloud/Documents/myfile.pdf.nextcloud`
Does download and launch the file successfully

**Comment by jeremyfritzen** [comment]
Hi!

Don't know if it's the right place to say it: I really love Nextcloud and job made by developers/contributors :)!

**Comment by antrmn** [comment]
I'm using Nextcloud 3.9.0 (Fedora 39, non flatpak). I've noticed that there's a row in `.config/Nextcloud/nextcloud.cfg` like this:

```
[General]
...
isVfsEnabled=false
```

Toggling this option seems to have no effect. Also, whenever I restart Nextcloud, it just gets reverted to `false`. What's the purpose of it?


> When I enable the experimental support that is there, it crashes when i hit the "enable virtual file support" button. The application segfaults and the last line in the log is this:
> 
> `2022-01-08 10:10:33:653 [ warning default unknown:0 ]: QObject::connect(QDialog, OCC::AccountSettings): invalid nullptr parameter `

@sinni800 did you manage to find a way to get it to work? I'm having the same issue.

**Comment by tnjoga** [comment]
> Toggling this option seems to have no effect. Also, whenever I restart Nextcloud, it just gets reverted to false. What's the purpose of it?


i have the same problem using Debian 12 and Nextcloud 3.9.0, i have also tried different package sources but the problem occurred with all of them.

**Comment by palto42** [comment]
For the Solus Linux packaging I had to move the plugin files `/usr/lib64/nextcloudsync_vfs_suffix.so` and `/usr/lib64/nextcloudsync_vfs_xattr.so`  to `/usr/lib64/qt5/plugins`.
This was reported as an issue for the AppImage build #3781, maybe worth to check the same on other distros?

**Comment by antrmn** [comment]
> For the Solus Linux packaging I had to move the plugin files `/usr/lib64/nextcloudsync_vfs_suffix.so` and `/usr/lib64/nextcloudsync_vfs_xattr.so` to `/usr/lib64/qt5/plugins`. This was reported as an issue for the AppImage build #3781, maybe worth to check the same on other distros?

In the Fedora version there's no `nextcloudsync_vfs_*.so` file.

In the end I've dropped it and picked up the AppImage. Now VFS with suffixes works. (No idea for `vfs_xattr`, though)

**Comment by palto42** [comment]
The `_vfs_` packages for Fedora are containe in the `devel` sub-package [nextcloud-client-devel-3.9.0-1.fc38.x86_64.rpm](https://fedora.pkgs.org/38/fedora-updates-x86_64/nextcloud-client-devel-3.9.0-1.fc38.x86_64.rpm.html), maybe because this is still an experimental feature?
They are installed in `/usr/lib64`, so maybe they need to be moved as I commented above.

**Comment by antrmn** [comment]
> The `_vfs_` packages for Fedora are containe in the `devel` sub-package [nextcloud-client-devel-3.9.0-1.fc38.x86_64.rpm](https://fedora.pkgs.org/38/fedora-updates-x86_64/nextcloud-client-devel-3.9.0-1.fc38.x86_64.rpm.html), maybe because this is still an experimental feature? They are installed in `/usr/lib64`, so maybe they need to be moved as I commented above.

Thank you for pointing it out. I installed the package, followed the steps above and can confirm that it is now possible to use the virtual files feature with the Fedora build of Nextcloud.

**Comment by Neshura87** [comment]
Fedora 38, Non Flatpak Version (3.9.0-1)

Installed nextcloud-client-devel, tried symlinking and copying `nextcloudsync_vfs_suffix.so` and `nextcloudsync_vfs_xattr.so` to `/usr/lib64/qt5/plugins/` but `isVfsEnabled` keeps resetting to false.
Am I missing something obvious here?
Do I also need to set `showExperimentalOptions = true` for it to work?

**Comment by afzl-wtu** [comment]
Any roadmap?

**Comment by mgallien** [comment]
having a proper working VFS implementation on linux is definitely a big amount of work needing support from current linux desktop implementation (think of files explorers, virtual files stack, ....)
gathering common interest on the feature from a broad set of people would help a lot

**Comment by mgallien** [comment]
`vfs_suffix` has way too much limitations to really want to use this as the solution even if all bugs were fixed

**Comment by KaeTuuN** [comment]
@mgallien The missing VFS implementation for Linux is one of just a few Blockers left for switching over to Linux as the primary Client OS. Is there any realistic guess on how much work it would be to implement it?
Maybe there are other companies out there willing to let one of their devs do the work or pay a dev for it!?

**Comment by tomassplatch** [comment]
> 

I can confirm that doing this after installing nextcloud-client-devel worked for me in Fedora.

**Comment by tomassplatch** [comment]
> For the Solus Linux packaging I had to move the plugin files `/usr/lib64/nextcloudsync_vfs_suffix.so` and `/usr/lib64/nextcloudsync_vfs_xattr.so` to `/usr/lib64/qt5/plugins`.

I can confirm that doing this after installing nextcloud-client-devel worked for me in Fedora.

**Comment by antrmn** [comment]
> FUSE is not a good solution when network is involved because the normal file API you end up using to access FUSE filesystems is not able to cope with network errors. 

Doesn't rclone use FUSE to mount remote storages? Are they settling with a flawed technology?

**Comment by r2evans** [comment]
@tobiasKaminsky I'm running Ubuntu 23.04, have `3.9.4-20230908.164710.c8fec0118-1.0~lunar1` installed, and with experimental features enabled, VFS is working for me: when I open the `.nextcloud` file (generally with `xdg-open`, optionally with nautilus), it is downloaded/renamed and opened correctly.

However, there are files I want to download and _not_ open. Often this is because I need to open it in a non-standard app, or want to pre-download so that I can reference the files later. (Examples: duckdb databases, data files to process with R or python, etc). The only way I've found to do this is to do a "full open" and then immediately exit the default application. In some cases this is just inconvenient, for others it might have consequences (large cpu/memory blooms, "on-open" macros in the file itself, etc).

I see your first post mentioned the suggestion of

> nextcloudcmd myVirtualImage.png.nextcloud -> downloads and opens it


In the documentation, in addition to the credentials required to connect, it looks as if it is expecting to sync to a different directory. Is it your intention with this plan to be able to do this inside a directory being serviced by the full-up `nextcloud` desktop app on individual files?

```
cd ~/my-nextcloud-syncdir/some-subdir/
nextcloudcmd myVirtualImage.png.nextcloud
```

and expect that it would download it locally? Perhaps with `--no-open` as an argument to short-circuit the default-app usage.

I've seen other comments _against_ implementing per-file selection of auto-sync in the app. Personally, I don't think the nextcloud desktop settings modal/app is the right place for individual files, I could easily see the `nextcloudcmd` (or some other companion utility) communicate with the full app to trigger some actions (such as make-available-locally).

**Comment by darkdragon-001** [comment]
Linking the discussion in https://gitlab.freedesktop.org/xdg/xdg-specs/-/issues/83 for visibility. I think a Freedesktop spec would be the right place for an implementation.

**Comment by mgallien** [comment]
Additional context about things happening related to this topic
https://www.volkerkrause.eu/2023/09/23/nextcloud-conference-2023.html
https://dragotin.codeberg.page/posts/openvfs/

**Comment by darkdragon-001** [comment]
[FUSE Passthrough Mode supported since Linux 6.9](https://www.phoronix.com/news/Linux-6.9-FUSE-Passthrough) could potentially solve the performance gap.

**Comment by ados8** [comment]
Really hoping Linux catches ground with Windows features.
I started transitioning Windows desktops to Linux and was shocked by lack of Nextcloud features.
It's generally recommended to run the server on the Linux platform, so I assumed it would function better on Linux desktop.
There's no virtual files, no thumbnail previews, no right click options etc.
I enabled the experimental option and it appeared to work, rebooted and then it starting syncing everything.
![Nextcloud](https://github.com/nextcloud/desktop/assets/49982433/dd34490a-5d3b-4f7e-bd4e-89bebfde3eb3)
Can see that virtual files is enabled, double checked the config but it was syncing everything.
Really want to daily drive Linux but might have to stick with Linux servers and Microsoft Windows.

**Comment by bluefoxconsultant** [comment]
Don't know whether this will help anyone, but I was able to configure my Nextcloud and Google Drive in a different way which I find works much better than using the Nextcloud Desktop (even with experimental features enabled).

1. Make sure rclone is installed on your machine
2. Run 'rclone configure' and configure your Nextcloud drive under the WebDAV section (Nextcloud is one of the choices of the menu that follows)
3. Create an empty folder to mount the rclone drive (in this case, mkdir /home/<user>/Nextcloud)
5. Run this or even make an sh file to run at startup:
`rclone mount  <nameofyourmount>: /home/<user>/Nextcloud --daemon --vfs-cache-mode full --vfs-cache-max-age 12h`

This will mount your Nextcloud (as configured in step 2) within the folder you've created. The cache mode is excellent and there are many more options for customizing the command itself.

**Comment by woiwoiwoiwoi** [comment]
I've been able to configure VFS on Debian without any issues, but on OpenSUSE Tumbleweed I ran into (mostly solvable) issues, as described in #6737.

@ados8 have you tried the AppImage? Also, for the right-click menu, a file explorer extension (the appropriate one for your desktop environment; also named differently on different distros) needs to be installed as well. But then it works well for me on both Debian and SUSE, both running Gnome.

**Comment by delete-your-github-today** [comment]
Please implement this as a FUSE filesystem. We need full compatibility with all VFS features. It's certainly possible to do this despite network delays/errors. rclone's FUSE support with its full caching mode is already effectively the functionality I want, it would just be nice to have this built-in to the nextcloud client.

**Comment by BloodyIron** [comment]
Not a current blocker for me but I foresee $futureMe caring about this significantly, as well as my clients. Here's hoping this can come to fruition and be generally available as a default ;)

**Comment by flamingm0e** [comment]
I still don't understand why this is still an issue.  Owncloud has it.  I'm running NC and OC in parallel and OC just has the better client overall.

**Comment by r2evans** [comment]
@flamingm0e have you experienced any issues running the OC desktop client against an NC server (or vice-versa)?

**Comment by flamingm0e** [comment]
> @flamingm0e have you experienced any issues running the OC desktop client against an NC server (or vice-versa)?

Sorry.  I didn't mean to insinuate I was running both clients against a single NC server.  I have both servers running:
NC AIO and OCIS.

**Comment by ados8** [comment]
> I still don't understand why this is still an issue. Owncloud has it. I'm running NC and OC in parallel and OC just has the better client overall.

Yeah it's a shame, open source software that ironically runs the best on proprietary commercial software that is Windows.
Further to the irony it appears Mac OS well get support first, for now we dream.

**Comment by Fuseteam** [comment]
well if OC has it, and the code is open source, then someone from the community could open a pull request to add to NC, no?

**Comment by Fuseteam** [comment]
hmmm as far as i can find owncloud uses placeholder files with a custom extension on mac and linux: https://github.com/owncloud/client/wiki/Virtual-Files

I presume that they replace this file once the user actually tries to open the file :thinking:

**Comment by antrmn** [comment]
> I presume that they replace this file once the user actually tries to open the file 🤔

That's basically how it currently works using the beta branch of Nextcloud Client on Linux.

A suitable example may be [onedriver](https://github.com/jstaf/onedriver.git) which basically uses FUSE, but as you can see in its README (see [Known Issues](https://github.com/jstaf/onedriver#known-issues--disclaimer)) it does not integrate well with file managers and has a few of rough edges, so I can't blame NC's developers for wanting a better solution.

Unfortunately it appears that no steps have been made in estabilishing a proper "Cloud Files API" for the Linux Desktop

**Comment by Fuseteam** [comment]
Sounds like we need an standard API on linux for it, tho i'm not sure is this would be portal or a protocol issue 🤔

**Comment by johnjaylward** [comment]
Would it make sense to use a project like [GDFS](https://github.com/robin-thomas/GDFS) as a baseline and implement a similar FUSE base filesystem that would work similarly? ... Not that I know if this would work or not. Just saw it may be an option

**Comment by jthoward64** [comment]
> Don't know whether this will help anyone, but I was able to configure my Nextcloud and Google Drive in a different way which I find works much better than using the Nextcloud Desktop (even with experimental features enabled).
> 
>     1. Make sure rclone is installed on your machine
> 
>     2. Run 'rclone configure' and configure your Nextcloud drive under the WebDAV section (Nextcloud is one of the choices of the menu that follows)
> 
>     3. Create an empty folder to mount the rclone drive (in this case, mkdir /home//Nextcloud)
> 
>     4. Run this or even make an sh file to run at startup:
>        `rclone mount  <nameofyourmount>: /home/<user>/Nextcloud --daemon --vfs-cache-mode full --vfs-cache-max-age 12h`
> 
> 
> This will mount your Nextcloud (as configured in step 2) within the folder you've created. The cache mode is excellent and there are many more options for customizing the command itself.

Honestly, I wonder if just adding an optional rclone dependency and automatically managing rclone mounts would be the way to do this for Linux

**Comment by Fuseteam** [comment]
that does look promising: https://rclone.org/commands/rclone_mount/#vfs-virtual-file-system
rclone even supports macos and windows too

**Comment by Lichtblitz** [comment]
rclone is an entirely different use case. It may fit many people and will not fit many more. For example rclone caches until the cache expires. It will not sync changes from the cloud during that time. If collaborating with others this may quickly become a deal breaker for offline availability. 
Secondly, there is no option for on demand sync. I have devices that wouldn't even fit my whole cloud storage if I wanted to. Plus the plethora of unnecessary syncing activity.
Microsoft implemented a close to perfect solution with the API they introduced directly in Windows. OneDrive uses the same API with the same great and seamless integration, that third parties use as well. Dropbox, Box, Google drive, etc.
Having a similar API baked into Linux without applications and users having to know about special files and providers would be awesome.

**Comment by phirestalker** [comment]
> rclone is an entirely different use case. It may fit many people and will not fit many more. For example rclone caches until the cache expires. It will not sync changes from the cloud during that time. If collaborating with others this may quickly become a deal breaker for offline availability. Secondly, there is no option for on demand sync. I have devices that wouldn't even fit my whole cloud storage if I wanted to. Plus the plethora of unnecessary syncing activity. Microsoft implemented a close to perfect solution with the API they introduced directly in Windows. OneDrive uses the same API with the same great and seamless integration, that third parties use as well. Dropbox, Box, Google drive, etc. Having a similar API baked into Linux without applications and users having to know about special files and providers would be awesome.

According to the documentation at https://rclone.org/commands/rclone_mount/#vfs-cache-mode-full it uses sparse files. This means it will not download the whole file and only the parts of it that are needed. You can also set a short caching time to save disk space. There is even a mode with no caching, although some applications can not work with this mode.

Cache expiration is configurable. It will even poll for changes at a configurable interval. You can make it 1s. On demand sync is not necessary as it is polling for changes and syncing them constantly. 

I am unsure of the syncing activity. Since the folks working on Linux do not seem too interested in creating a universal VFS implementation, rclone can be a good alternative. I would be grateful if Nextcloud would consider using rclone in the background until Linux implements the necessary API.

Currently, I use the built-in WebDAV functionality in KDE. Unfortunately, this is not available in any non-KDE apps or from the terminal. I would have used rclone if this functionality was missing from KDE.

**Comment by Fuseteam** [comment]
> Secondly, there is no option for on demand sync. I have devices that wouldn't even fit my whole cloud storage if I wanted to. 

that is actually what vfs-cache-mode full is with rclone, it does **not** download the actual the data, just the metadata needed to download the data on demand

**Comment by lachlanshoesmith** [comment]
> When I enable the experimental support that is there, it crashes when i hit the "enable virtual file support" button. The application segfaults and the last line in the log is this:
> 
> `2022-01-08 10:10:33:653 [ warning default unknown:0 ]: QObject::connect(QDialog, OCC::AccountSettings): invalid nullptr parameter `

did anyone end up figuring out what causes this? having this problem on 3.14.0daily on my laptop, dont on my similarly-setup desktop for some reason. running arch with hyprland fwiw, can reproduce using the same steps of adding `enableExperimentalOptions=true` to `nextcloud.cfg` and then trying to enable virtual files support in the client

**Comment by antrmn** [comment]
Maybe openVFS could be considered as a project proposal for initiatives like Outreachy Internships or Google Summer of Code? GNOME was selecting project proposals for Outreachy just a month ago.

 This could provide an opportunity for [openVFS](https://github.com/dragotin/openvfs/discussions/1) to be explored further

**Comment by thibaultmol** [comment]
I'm also kind of confused why the current virtual file system implementation that you can enable using `showExperimentalOptions` is still so hidden away. If there are bugs, then what are the bug reports/issue numbers that are blocking it from being rolled as a stable feature?

(I think some of the above mentioned solutions would also be nice, but i'm mainly wondering why the current implementation hasn't shipped as stable for years now)

**Comment by bluefoxconsultant** [comment]
Personally I've never been able to make the experimental VFS work. 
It created placeholder files as I expected it to, however upon trying to open those files, nothing happened (tried through the Open menu in applications as well as launching the file directly from Dolphin).
I've found rclone to generally be a vastly superior solution, although there are some slowdowns when using Google Drive (especially when listing directories and files).

**Comment by woiwoiwoiwoi** [comment]
For me, it's been working flawlessly for 2-3 years on Debian and OpenSUSE Tumbleweed. I've also wondered why it's still so hidden away.

**Comment by langfingaz** [comment]
> For me, it's been working flawlessly for 2-3 years on Debian and OpenSUSE Tumbleweed. I've also wondered why it's still so hidden away.

What desktop environment are you using? I assume the support varies between different DEs and if one is using QT, GTK or other applications.

**Comment by Fuseteam** [comment]
> For me, it's been working flawlessly for 2-3 years on Debian and OpenSUSE Tumbleweed. I've also wondered why it's still so hidden away.

i believe the comment above yours answers that question

**Comment by r2evans** [comment]
> It created placeholder files as I expected it to, however upon trying to open those files, nothing happened (tried through the Open menu in applications as well as launching the file directly from Dolphin).

This _might_ be due to something external to NC. I suggest looking into how default-applications are defined in your distro to see if `.nextcloud` is correctly defined. I don't know how most distros do it, but after `strace`-ing `mimeopen .nextcloud`, I found it looking at `/usr/share/mime/globs` which includes

```
...
image/x-3ds:*.3ds
audio/x-ms-wma:*.wma
application/vnd.nextcloud:*.nextcloud
text/x-python3:*.py
application/vnd.lotus-1-2-3:*.123
...
```

which should match the placeholder with the associated mime type `application/vnd.nextcloud`. That type is matched to an executable via `/usr/share/applications/mimeinfo.cache` (or a `~/.local/share/applications` or several other places to find multiple `mimeinfo.cache` files), where

```
...
application/vnd.ms-wpl=vlc.desktop;
application/vnd.ms-xpsdocument=org.gnome.Evince.desktop;
application/vnd.nextcloud=com.nextcloud.desktopclient.nextcloud.desktop;
application/vnd.oasis.opendocument.base=libreoffice-base.desktop;
application/vnd.oasis.opendocument.chart=libreoffice-calc.desktop;
```

And that points to `/usr/share/applications/com.nextcloud.desktopclient.nextcloud.desktop` (ditto on `~/.local/share/...`), which ultimately points to the `Exec`utable that is called when something tries to open a `*.nextcloud` file.

```
[Desktop Entry]
Categories=Utility;X-SuSE-SyncUtility;
Type=Application
Exec=nextcloud %u
Name=Nextcloud Desktop
Comment=Nextcloud desktop synchronization client
GenericName=Folder Sync
Icon=Nextcloud
Keywords=Nextcloud;syncing;file;sharing;
X-GNOME-Autostart-Delay=3
MimeType=application/vnd.nextcloud;x-scheme-handler/nc;
Actions=Quit;
```

(Long story, thanks for walking all of it with me.)

So if you can verify that all of that is correctly defined, and that something like `mimeopen somefile.docx.nextcloud` works up until the point that `nextcloud %s` is attempted, then the problem is certainly with the NC desktop utility. If it doesn't get that far, then the linux-hack for VFS (`.nextcloud` file extension) may not be why it is not working on your system.

**Comment by woiwoiwoiwoi** [comment]
> What desktop environment are you using? I assume the support varies between different DEs and if one is using QT, GTK or other applications.

Gnome.

**Comment by MrRinkana** [comment]
A workaround for KDE users is adding the nextcloud account to dolphin (system settings -> accounts) as that provides a way to get "virtual" files, although it seems to work more akin to a network share, such that files always get downloaded (possibly some temp. caching) rather than being stored locally after opening them once. It has some benefits like easily uploading folders etc. This is what I do since:

For me the nc-client just crashes when trying to enable the experimantal-VFS, I don't see any syck option during install or adding a sync connection, I can only add it afterwards by clicking the three dots on an existing sync-connection (and then it crashes).

**Comment by antrmn** [comment]
> Maybe openVFS could be considered as a project proposal for initiatives like Outreachy Internships or Google Summer of Code? GNOME was selecting project proposals for Outreachy just a month ago.
> 
> This could provide an opportunity for [openVFS](https://github.com/dragotin/openvfs/discussions/1) to be explored further

I reiterate my previous suggestion, this time bringing into consideration the following article: [GNOME STF 2025: Call for Ideas and Contractors](https://discourse.gnome.org/t/gnome-stf-2025-call-for-ideas-and-contractors/25478)

 
> We’re currently putting together a Sovereign Tech Fund application for next year, both collecting ideas for projects and investigating what we’ll be able to staff. We’ve already got a list of focus areas and ideas that fit, and we’re in the process of having meetings with various teams of maintainers across the project. However, we also want to open this up to the wider community.
> 
> Our focus areas are:
> 
>     [...]
>     Platform features in GNOME, Freedesktop, Flatpak, portals, and related projects. For instance: xdg-intents, various Flatpak portal improvements, session save/restore, notifications, doze/wakeup/power management, search, text-to-speech, autofill, etc.
>     [...]
> 
> If you have any ideas in these areas, and/or would be available to work on them on a contract basis, please contact us ([stf@gnome.org](mailto:stf@gnome.org)), or leave a comment below.

**Comment by plauderer** [comment]
Any experience on enabling the feature on Linux Mint? Is there a chance to get VFS support on Linux available in the future without need to enable experimental features?

**Comment by EKCJ** [comment]
> Any experience on enabling the feature on Linux Mint? Is there a chance to get VFS support on Linux available in the future without need to enable experimental features?

I've been using VFS on Mint for several months, it's been working well.

As of an hour ago it appears to be bugging out and is doing a full sync of everything regardless of config, which is what prompted me to check back in on this issue to see if anyone else is having the same problem - but I haven't done any troubleshooting yet, so no idea what's causing it, and it shouldn't put you off trying the feature on Mint.

**Comment by Shardboii** [comment]
So i've found out the steps on how to get VFS working most of the times. This thread has been the most helpful with this problem.
Im using Fedora 42 and have tried it on multiple systems with KDE Plasma.

1. Install Nextcloud using DNF (or via your Package Manager, i don't know if the following steps will work tho)
```
sudo dnf in -y nextcloud-client nextcloud-client-devel nextcloud-client-libs nextcloud-client-nautilus #or dolphin
```
Nextcloud-Devel contains VFS libs as mentioned earlier in this thread
2. Check the Version, it should say QT6 or QT5 somewhere. Remember that
3. Move Nextcloud VFS qt plugins to the right folder
```
sudo mv /usr/lib64/nextcloudsync_vfs_suffix.so /usr/lib64/qt5/plugins/
```
Depending on what QT version you have you should either move it to QT5 or QT6
i did QT6
4. edit ~/.config/Nextcloud/nextcloud.cfg and add showExperimentalOptions=true before you start the client. Make sure the Config file is fully empty except these two Lines
```
[General]
showExperimentalOptions=true
```
If all goes well, you should see the VFS option during the inital setup where you get to decide how you want the files to be stored.

<img width="505" height="37" alt="Image" src="https://github.com/user-attachments/assets/fbacd9af-b540-4f68-a7f5-e69f277d97db" />

**Comment by PichlAlex** [comment]
VFS would be a really great feature - especially as i have way more stored on my nextcloud as my Laptop has free disk space available.

my setup that works for me:
- davfs for access to everything, but only usable when internet is available. I wrote a small systemd script that checks if i can connect to the nextcloud, then connects davfs... also handles some error cases (still optimizing that). I increased the local cache in davfs to get most accessed files with higher performance
- most important folders are synced using the nextcloud client that i may need offline available.

i usally work with davfs

**Comment by woiwoiwoiwoi** [comment]
@PichlAlex Why don't you just use VFS then? I've been using it with only minor (and solvable) hiccups for close to 5 years now on both Debian and on OpenSUSE Tumbleweed.

**Comment by Murazaki** [comment]
> > It created placeholder files as I expected it to, however upon trying to open those files, nothing happened (tried through the Open menu in applications as well as launching the file directly from Dolphin).
> 
> [@bluefoxconsultant](https://github.com/bluefoxconsultant)
> 
> This _might_ be due to something external to NC. I suggest looking into how default-applications are defined in your distro to see if `.nextcloud` is correctly defined. I don't know how most distros do it, but after `strace`-ing `mimeopen .nextcloud`, I found it looking at `/usr/share/mime/globs` which includes
> 
> ```
> ...
> image/x-3ds:*.3ds
> audio/x-ms-wma:*.wma
> application/vnd.nextcloud:*.nextcloud
> text/x-python3:*.py
> application/vnd.lotus-1-2-3:*.123
> ...
> ```
> 
> which should match the placeholder with the associated mime type `application/vnd.nextcloud`. That type is matched to an executable via `/usr/share/applications/mimeinfo.cache` (or a `~/.local/share/applications` or several other places to find multiple `mimeinfo.cache` files), where
> 
> ```
> ...
> application/vnd.ms-wpl=vlc.desktop;
> application/vnd.ms-xpsdocument=org.gnome.Evince.desktop;
> application/vnd.nextcloud=com.nextcloud.desktopclient.nextcloud.desktop;
> application/vnd.oasis.opendocument.base=libreoffice-base.desktop;
> application/vnd.oasis.opendocument.chart=libreoffice-calc.desktop;
> ```
> 
> And that points to `/usr/share/applications/com.nextcloud.desktopclient.nextcloud.desktop` (ditto on `~/.local/share/...`), which ultimately points to the `Exec`utable that is called when something tries to open a `*.nextcloud` file.
> 
> ```
> [Desktop Entry]
> Categories=Utility;X-SuSE-SyncUtility;
> Type=Application
> Exec=nextcloud %u
> Name=Nextcloud Desktop
> Comment=Nextcloud desktop synchronization client
> GenericName=Folder Sync
> Icon=Nextcloud
> Keywords=Nextcloud;syncing;file;sharing;
> X-GNOME-Autostart-Delay=3
> MimeType=application/vnd.nextcloud;x-scheme-handler/nc;
> Actions=Quit;
> ```
> 
> (Long story, thanks for walking all of it with me.)
> 
> So if you can verify that all of that is correctly defined, and that something like `mimeopen somefile.docx.nextcloud` works up until the point that `nextcloud %s` is attempted, then the problem is certainly with the NC desktop utility. If it doesn't get that far, then the linux-hack for VFS (`.nextcloud` file extension) may not be why it is not working on your system.

Thanks for the info on the mime type, that was not set up properly on my NixOS with KDE desktop OS. Do you happen to know if there's a way it could show previews on Dolphin ?

**Comment by PichlAlex** [comment]
> [@PichlAlex](https://github.com/PichlAlex) Why don't you just use VFS then? I've been using it with only minor (and solvable) hiccups for close to 5 years now on both Debian and on OpenSUSE Tumbleweed.

@woiwoiwoiwoi 
good question... i installed this in a ubuntu 16 LTS VM and upgraded it since.... and never bothered to change this, because it was just working enough for me.
... and a little bit of "experimental" i try to avoid for this kind of things when i have to rely on to "just work".

i changed it now, and wow - it works great (for the last days)

**Comment by dbrand666** [comment]
For all of you who are claiming it works...

If I go to the Nextcloud directory in a shell, do I see files I can just use or do I see placeholder files named something.nextcloud? When I last tried it on Fedora (a few months ago), I saw placeholder files and couldn't get much shell work done.

Compare this to rclone mounts which are pretty much completely transparent.

Nextcloud:
\- Tricky to enable experimental feature.
\+ Transparent file access in file explorers (with helpers).
\+ Keeps local cached copy with bidirectional sync (I think?).
\+ Offline access to file that happen to be cached.

Rclone:
\- Configuration is a bit challenging to the uninitiated.
\+ Completely transparent file access from both file explorer and shell.
\- Fiddly caching options.
\- No offline access.

So I'm running a little of each. Nextcloud without VFS for a few directories I need even when I'm offline and Rclone for the rest.

I'm looking forward to the day that rclone mount supports proper caching and bidirectional sync. I wouldn't mind if Nextcloud surprises us with a proper FUSE implementation though.

**Comment by woiwoiwoiwoi** [comment]
Well, "claiming" sounds a bit accusatory. The question is does it work for you or me in your or my context. My context for files synced from my Nextcloud server is 100% file explorer access via Nautilus on Gnome. Yes, the virtual files are `filename.extension.nextcloud`, which is not perfect. But for my personal use case, that is completely irrelevant.

Offline access does not only work for files that happened to be cached, but you can choose a file or folder, right-click on it and set it to "always available offline", which works very reliably. The only issue here (for me) is that the little icon next to the file does not change correctly - it always stays at "available offline", whether the file is actually available or not. But then again, I only need to look at the file extension: if it's `.nextlcoud` then it's not available, if it's the actual extension then it's available. I simply ignore those icons.

So yes, the current implementation is not perfect and probably not for every use case. But the actual job of syncing only files that are required has been working for me for over 5 years, with zero data loss or corruption. And it allows me to access 1 TB of files on my server, for instance, from laptops or VMs with small hard drives, without having to filter, choose, or put any manual labour into it.

**Comment by Murazaki** [comment]
> Well, "claiming" sounds a bit accusatory. The question is does it work for you or me in your or my context. My context for files synced from my Nextcloud server is 100% file explorer access via Nautilus on Gnome. Yes, the virtual files are `filename.extension.nextcloud`, which is not perfect. But for my personal use case, that is completely irrelevant.
> 
> Offline access does not only work for files that happened to be cached, but you can choose a folder, right-click on it and set it to "always available offline", which works very reliably. The only issue here (for me) is that the little icon next to the file does not change correctly - it always stays at "available offline", whether the file is actually available or not. But then again, I only need to look at the file extension: if it's `.nextlcoud` then it's not available, if it's the actual extension then it's available. I simply ignore those icons.
> 
> So yes, the current implementation is not perfect and probably not for every use case. But the actual job of syncing only files that are required has been working for me for over 5 years, with zero data loss or corruption. And it allows me to access 1 TB of files on my server, for instance, from laptops or VMs with small hard drives, without having to filter, choose, or put any manual labour into it.

Honestly it does work well if you ask nextcloud to open .nextcloud files, so my only issue is on icons and thumbnails. I'm pretty sure this could be handled as complementary metadata on the .nextcloud file though.

**Comment by woiwoiwoiwoi** [comment]
> Honestly it does work well if you ask nextcloud to open .nextcloud files

Oh yes, absolutely, should have made that clearer. You launch any virtual file you want and it opens in the appropriate application. As a simple user you don't notice that the file is actually opened by the Nextcloud client, downloaded in the background, and then passed on to whichever application is the appropriate one.

**Comment by MrRinkana** [comment]
For Tumbleweed @Murazaki's process translates to:

1. Install nextcloud-desktop-vfs-plugin
2. Edit .config/Nextcloud/nextcloud.cfg to add "showExperimentalOptions=true"
3. Enable vfs when setting up a folder sync (I had to do it twice by right-clicking the folder-share afterwards) - you are looking for the purple "Virtual file support enabled"

No need to move any files!

Thumbnails and browsing in cli is still not working as noted above but that needs a common standard in the linux ecosystem to be doable..

**Comment by dbrand666** [comment]
> FUSE is not a good solution when network is involved because the normal file API you end up using to access FUSE filesystems is not able to cope with network errors.

Is this still true? Was it ever true?

FUSE has been used to implement NFS, SSH, and various cloud storage mounts.

**Comment by PichlAlex** [comment]
nextcloud-client-4.0.7-2.fc43 is available for installing as rpm

many thanks to 2026-03-17	Mukundan Ragavan 

now it is only:

- installing the rpm via Discover
- adding ~/.config/Nextcloud/nextcloud.cfg:
```
[General]
showExperimentalOptions=true
```
- add the sync folder
- "activate VFS"

extensions in Dolphin works out of the box!

**Comment by dbrand666** [comment]
The "activate VFS" step doesn't so much work as crash, but it's progress of a sort.

**Comment by WerWolv** [comment]
Doing some debugging, it looks like the crash happens here https://github.com/nextcloud/desktop/blob/c4f7d3dc95e0d0e7f80f2128ed244c682cb1d5b3/src/gui/wizard/owncloudwizard.cpp#L539. `msgBox` is `nullptr` because `bestVfsMode` turned out to be `Vfs::Off` since some plugin isn't available. I guess the one that should be used on Linux would be the `WithSuffix` one since `XAttr` also leads to that same outcome

Edit:
Got it working. For some reason, on Fedora the required plugins are only part of the -devel package. So you need to install the `nextcloud-client-devel` package as well as the `nextcloud-client` package.
After that it still doesn't work though because the plugin load path seems to be wrong. Manually setting the `QT_PLUGIN_PATH` env var to the right place makes it work for me.

```
$ sudo dnf install nextcloud-client nextcloud-client-devel
$ QT_PLUGIN_PATH=/usr/lib64/ nextcloud
```

**Comment by fhaefemeier** [comment]
Seems to work as expected (test it since a few hours). Is there a way to delete the local copy but not delete the remote version?
I use Fedora Linux Mint (Nemo file browser).

Edit:
Ok, I answered by myself. There is a "Nextcloud" menu, but only if Nemo is started with the env variable `QT_PLUGIN_PATH` mentioned by @WerWolv . Additional the icons show now a sync status.

**Comment by AlexHaan-i** [comment]
I also seem to be able to try this with the beta channel [AppImage Nextcloud Desktop Client Version 33.0.2 (Ubuntu running on xcb)] on my Ubuntu 22.04.
Then after adding that config setting and restarting I can enable VFS.

But after 100% on one CPU core for about an hour nothing has changed. The Nextcloud client was still running. But CPU usage went back to normal. I guess it was scanning the whole Nextcloud share first. But there are many thousands of files on there.

Creating .nextcloud files for all of those files seems excessive. Do the solutions on the other file systems work the same way?

**Comment by vasyugan** [comment]
> > FUSE is not a good solution when network is involved because the normal file API you end up using to access FUSE filesystems is not able to cope with network errors.
> 
> Is this still true? Was it ever true?
> 
> FUSE has been used to implement NFS, SSH, and various cloud storage mounts.

Interesting point! What does e.g. sshfs do when the network connection goes away? Does it have some form of caching? To me it seems, improving the reliability of FUSE in such scenarios, so that it supports on-demand downloading and handles situations where the network is inaccessible,  would be a more sustainable solution that Nextcloud reinventing the wheel in a manner that only benefits the Nextcloud client and only works in certain environments (I don't think that .nextcloud files are useful e.g. in the terminal, are they?). The libfuse repository is [here](https://github.com/libfuse/libfuse). I've not found an issue that seems related there, yet. But I might not have used the right search terms. I'm also not sure what capabilities FUSE is actually lacking. 

But the real solution would probably be something like [libcloudproviders](https://gitlab.gnome.org/GNOME/libcloudproviders). Alas, the project is dormant, last release in 2019.

**Comment by Fuseteam** [comment]
wonder if this is relevant for us https://git.kernel.org/pub/scm/linux/kernel/git/vfs/vfs.git/

**Comment by luciandf** [comment]
> What does e.g. sshfs do when the network connection goes away?

i am not a programmer but i am really looking forward for VFS to be implemented for linux. having said that, why does it matter what sshfs does when the network connection id down? the nextcloud client is down too.

I have used sshfs to mount remote folders for years and it has two timeout variables that can be set to control how many times it tries to reconnect and also a feature to keep the connection alive. but I don't see why this is relevant to VFS. The desktop client should just "see" what files are in the user account which are not locally downloaded. Once the user downloads them then it is business as usual. Sorry about my naive comment, I am sure it doesn't help but I am really looking forward to VFS in linux!

**Comment by dbrand666** [comment]
> But the real solution would probably be something like [libcloudproviders](https://gitlab.gnome.org/GNOME/libcloudproviders). Alas, the project is dormant, last release in 2019.

Probably not. It's a Gnome DBUS project and wouldn't provide native transparent access either.

**Comment by sinni800** [comment]
It kind of sucks that Windows theoretically lapped linux here because they just enforced a single local api for "remote files". I feel like we need a standards push. I wonder actually if this would be an interesting way of applying portals. Intent of accessing files would have to be clear because otherwise it would just download everything on the first FS scan. It would be a maybe fun emulation of android's file access intent. This is just an idea though.

But I see seafile (with seadrive) be pretty well off with just FUSE instead of this marker file thing. They just have a local cache directory and will deny file access if it's offline. One could then let the user define if files are writeable when offline to make sure online locking works and such.

**Comment by Fuseteam** [comment]
> It kind of sucks that Windows theoretically lapped linux here because they just enforced a single local api for "remote files". 

This is because windows is 1 Operating System, while Linux is a family of operating systems.

Every protocol has it's pros and cons, the nextcloud client could utilize one specific protocol that suits their usecase and other linux operating systems can then adopt this protocol.

Most of the larger linux operating systems are aware of each other and collaborate behind the scenes on what they all should support— there are time they don't agree and don't implement what others do implement (cough gnome) but that can ultimately be driven by users applications do and do not support. So getting in touch with one, will eventually pan out to the rest. 

That said ultimately it comes down to desktop enviroments, of which the bigger ones are gnome or kde....of which kde is probably the  more flexible one

**Comment by troed** [comment]
> But I see seafile (with seadrive) be pretty well off with just FUSE instead of this marker file thing. They just have a local cache directory and will deny file access if it's offline. 

I used Seafile for years before switching to Nextcloud and I'm surprised that a similar solution hasn't been implemented here. It worked just fine.

"Perfect is the enemy of good" seems to apply.

**Comment by Fuseteam** [comment]
> But I see seafile (with seadrive) be pretty well off with just FUSE instead of this marker file thing. They just have a local cache directory and will deny file access if it's offline. One could then let the user define if files are writeable when offline to make sure online locking works and such.

Tho i think the goal of VFS is to not save everything, but only the most recently accessed. It kinda beats the purpose if it downloads 10G in cache when it really only 1G is currently accessed

**Comment by troed** [comment]
> I'm surprised that a similar solution hasn't been implemented here. It worked just fine.

For information: I bit the bullet and started on such an implementation for Nextcloud. It's currently in open testing phase but _not_ where I would want anyone to run it with actual live data just yet. Thus I'm not linking it - anyone willing to put in the effort to find it is likely also going to understand not to risk their files with it :D 

LRU cache eviction, "mount and forget", no changes needed to Nextcloud.

**Comment by mgallien** [comment]
in case some of you may be interested, there is this project
https://github.com/opencloud-eu/openvfs
nothing ready yet in the Nextcloud desktop files client
contributions are very much welcome

**Comment by Plinsboorg** [comment]
+1 for this feature

**Comment by cdhermann** [comment]
Since virtual file support and on-demand sync are central topics in this issue, this alternative FUSE-based implementation might be of interest:

**Wusel** — A Nextcloud FUSE client was initially developed for internal company use and is now being made available as open source (Apache-2.0). 

It is a Nextcloud virtual filesystem written in Rust and mounted over FUSE. Server contents appear instantly, file contents are fetched only when opened, and specific files can be pinned to be kept offline.

→ https://github.com/itbh-at/wusel

Feedback and pull requests are welcomed. The current scope is limited to Linux and Nextcloud. Desktop integration is provided for GNOME/Nautilus, and a ready-made Fedora RPM is available. For other environments, the client must be built from source.

**Comment by troed** [comment]
> It is a Nextcloud virtual filesystem written in Rust and mounted over FUSE. Server contents appear instantly, file contents are fetched only when opened, and specific files can be pinned to be kept offline.
 
Thank you! I've been working on the same (Rust+FUSE on Linux) and I've not dared share the link publically until I've tested it back and forth :D Will test yours. Being able to successfully git clone into the vfs gave me the worst issues.

**Comment by cdhermann** [comment]
> tested it back and forth

Extensive internal testing was conducted, though no guarantees can be made. An initial single-threaded version worked adequately but rather slowly. The multi-threaded client now available is significantly faster. Bugs are naturally to be expected – as is typical for software – but the client has been thoroughly challenged in daily use. However, it must be noted that a certain degree of operational tunnel vision might exist, as only our specific usage patterns were evaluated.

**Comment by rgon** [comment]
As an alternative, and similar to @cdhermann 's version, I have just open-sourced the alternative Nextcloud desktop client w/VFS I have been working for a while:

https://github.com/rgon/ncrsDesktop

It includes a full tauri-based nextcloud desktop app replacement, appearing as a small tray icon with Login Flow v2 and implementing dashboard apps (for now, GNOME **Calendar/contacts sync**, a dash search provider, a cross-app search, also shows **notifications** and integrates with **NC Passwords** for quick team-shared password usage). It's modular so adding new apps should be easy. The sync service may also be run headlessly, without the tray app, if needed. It should be lightweight, though (~90MB RAM usage with the tray icon closed).

The VFS client is quite advanced: connects to the nextcloud **high-performance backend** for realtime updates. It allows streaming reads & chunked uploads, HTTP/3 transport, file & **directory list caching** (+ by default a zero-copy FUSE_PASSTHROUGH cache), prefetching, ETAG-based updates, has multithreaded requests and is generally optimized for performance and lots of large files (media & many pdfs), achieving >350mbps uploads over the internet on our (not ideal) deployment, 4x faster than the web ui (in our specific edge case - no quantitative guarantees but it'll be faster than the official clients for sure). 

The auto-installed file browser plugin also enhances: uses the server's **thumbnailer** directly exposing it to nautilus (even working for raw image files) and loading quite fast, shows sync/upload/cache/downloaded status, who owns/shared the file and respects groups/share **pemissions**, 'keep locally' for **offline** use,It also adds a 'show in browser' context menu option which eases the creation of public share links or conversation, from the files found in the desktop.

It's a long feature list but it's what we needed to keep the marketing people in our team happy 😆 

It's in production for us, we've been using it exclusively and are happy with it, thus why I open sourced it now. Happy to hear feedback!

**Comment by cdhermann** [comment]
> As an alternative, and similar to [@cdhermann](https://github.com/cdhermann) 's version, I have just open-sourced the alternative Nextcloud desktop client w/VFS I have been working for a while:

@rgon Great to see another Rust VFS client, and congrats on open-sourcing it. This thread was rather quiet for years, and now there are two independent on-demand VFS clients within approximately a month — that's the strength of free software, and more choice here is only good. The server thumbnails in the file manager (RAW included!) and the QUIC transport are nice touches Wusel doesn't have today.

For anyone comparing the two, a few things Wusel already does, so the picture is fair: it connects to the High-Performance Backend (notify_push) for instant cache updates, does chunked uploads, and serves the requested byte range live while it downloads the rest of the file in the background — so a large video starts playing right away, without a full download first. The app password is stored in the OS keyring, share permissions map to file permissions, and the cache is pruned by size and age. Each account is its own mount and its own systemd user service, so it also runs headless, with nobody logged in.

On packaging it goes fairly wide: a signed repository via the openSUSE Build Service for Fedora, openSUSE Tumbleweed, Debian 13 and Ubuntu 26.04 (x86_64 and aarch64), so it is GPG-verified and updates through the normal package manager — plus an AUR PKGBUILD for Arch and single RPM/.deb files per release. There is also an experimental macOS File Provider frontend, with Windows and KDE planned.

The main difference is scope, not the core engine — and here Wusel is the odd one out of the three. Both your client and the official on-demand client being worked on in this thread present as a tray application with extra built-in apps; Wusel deliberately ships neither and leans on the desktop's own tools instead — per-file status lives on the files as emblems and in the desktop's notifications. So your bundled GUI, the calendar/contacts and NC Passwords integration, and HTTP/3 are things Wusel doesn't do. The FUSE passthrough cache is a different case: it fits our model well — fetch a file once, then let the kernel serve it straight from disk — and it's something we'll likely pick up.

Different priorities, same goal. Happy to compare notes!

**Comment by rgon** [comment]
@cdhermann absolutely, the beauty of free software!

Absolutely agree, in our case it definitely has to be used by non-technical people, thus why all the bells and whistles.

However, I built the VFS-part as a completely separate program, which is in fact run separately via systemd and just commands a socket api to the frontend, in a similar way that the nautilus client calls it. My reasoning to split it was surely the same as yours when you decided to build a headless version. It indeed can be used separately, the desktop app is just a nice-to-have way to extend nextcloud (server) apps onto the desktop.

 IMO there's no reason why nextcloud passwords shouldn't just be a js-native app within the desktop client as well (same stack as the web version but miles better UX). That + setting the user status, which in a team may be useful, warrants the dashboard app.

Had I known your project a few moths back I'd just have based mine on it to be honest. I believe they can be parts of the same engine In fact, if you're up to it, I'd enjoy it if both projects converged. My GUI could speak the same API as both the backends, and the nautilus clients could also do the same, but they'd essentially end up converging in feature set and edge case fixes (some weird ones I found were overlaying .trackerignore to prevent server spam, faking a response given a read fingerprint to prevent the nautilus thumbnailer, creating a local partial directory tree index & keeping it updated with the etags...) and you probably also found some edge case as well. I added a CI testing suite with a webdav server, but it could of course be improved, and I'm working around the system sometimes - better clarification to GNOME (with x-gvfs) that this is a remote fs would clear things up, but that has to go on their team. 

My next concern in my project (as appears in the long TODO in my repo) is related to nextcloud auto-login for users that logged in with OAuth to our ubuntu laptops, so all their Nextcloud files, calendars, contacts, passwords... are already set up (essentially what kerberos is supposed to do but simpler and safer). This is a very enterprise-y use case that depends on me finishing that part, but this would be a use case that probably interests the NC org more than anyone else.

The best end-result for the community is probably to extend all this code base to a few common DEs/OSs and probably to get Nextcloud to integrate this effort as theirs.  I really believe that this desktop integration was a big hinderance to NC's usability especially on enthusiasts and businesses, and both our projects could be a good base for it. 

Let me know if you're up to collaborating on this, and if anyone here would like to join!

**Comment by camilasan** [comment]
I think the beauty of free and open software is collaboration (unless the project is not open to that...) and it is not in creating one tool for each unique use case. You are all going to bump into the same issues we bumped into years ago and since coding is so cheap now, you will probably just create a new project or just re-write the whole thing from scratch again. This is not about open source anymore.

**Comment by mgallien** [comment]
this is now done and available in master branch and future releases
https://github.com/nextcloud/desktop/pull/10635

**Comment by luciandf** [comment]
so this is it? VFS is coming to the official nextcloud linux desktop client? 🥳

**Comment by szaimen** [comment]
> so this is it? VFS is coming to the official nextcloud linux desktop client? 🥳

Yes. See https://github.com/nextcloud/desktop/pull/10635
